static/app/components/modals/importDashboardFromFileModal.tsx

```diff
@@ -0,0 +1,106 @@
+import {useState} from 'react';
+import {browserHistory} from 'react-router';
+
+import {createDashboard} from 'sentry/actionCreators/dashboards';
+import {addSuccessMessage} from 'sentry/actionCreators/indicator';
+import {Button} from 'sentry/components/button';
+import {CodeSnippet} from 'sentry/components/codeSnippet';
+import {t} from 'sentry/locale';
+import {normalizeUrl} from 'sentry/utils/withDomainRequired';
+import {
+  assignDefaultLayout,
+  getInitialColumnDepths,
+} from 'sentry/views/dashboards/layoutUtils';
+
+function ImportDashboardFromFileModal({organization, api, location}) {
+  const [file, setFile] = useState<File>();
+  const [dashboardData, setDashboardData] = useState('');
+  const [validated, setValidated] = useState(false);
+  const [error, setError] = useState();
+
+  function validateFile(fileToUpload) {
+    // todo(?): validate contents of JSON?
+
+    if (!(fileToUpload.type === 'application/json')) {
+      setError('The uploaded file must be JSON');
+      setValidated(false);
+      return false;
+    }
+    setValidated(true);
+    return true;
+  }
+
+  const handleFileChange = (e: ChangeEvent<HTMLInputElement>) => {
+    if (e.target.files) {
+      const fileToUpload = e.target.files[0];
+      if (validateFile(fileToUpload)) {
+        const fileReader = new FileReader();
+        fileReader.readAsText(fileToUpload, 'UTF-8');
+        fileReader.onload = event => {
+          const parsed = JSON.parse(event.target.result);
+          const formatted = JSON.stringify(parsed, null, '\t');
+          setDashboardData(formatted);
+          // setDashboardData(e.target.result);
+
+          // setDashboardData(JSON.stringify(e.target.result, null, '\t'));
+          // console.log(typeOf(JSON.stringify(e.target.result, null, '\t')));
+        };
+        // console.log(JSON.parse(data));
+        setFile(fileToUpload);
+      }
+    }
+  };
+
+  const handleUploadClick = async () => {
+    const dashboard = JSON.parse(dashboardData);
+    if (!file) {
+      return;
+    }
+
+    const newDashboard = await createDashboard(
+      api,
+      organization.slug,
+      {
+        ...dashboard,
+        widgets: assignDefaultLayout(dashboard.widgets, getInitialColumnDepths()),
+      },
+      true
+    );
+
+    addSuccessMessage(`${dashboard.title} dashboard template successfully added`);
+    loadDashboard(newDashboard.id);
+    // console.log('todo: upload the file in this function');
+  };
+
+  const loadDashboard = (dashboardId: string) => {
```

Comment:
This is copied from [another file](https://github.com/getsentry/sentry/blob/7c9acdc051c5251fa1535635e2413ec26c30b470/static/app/views/dashboards/manage/index.tsx#L247-L255). Since it's only used in one other place I'm not too concerned about needing to extract it to a higher order component, but wanted to flag this as a question during review